### PR TITLE
[SPARK-39714][PYTHON] Try to fix the mypy annotation tests

### PR DESCRIPTION
**Pull request by bzhaoopenstack**
This patch is for resolving the mypy annotation tests. For making the
 annotation tests, we figure out all failure tests and fix them.

### What changes were proposed in this pull request?

The changed files are all the errors raised when we execute dev/lint-python. All is related mypy annotation test failure.

### Why are the changes needed?

We need to resolve them for avoiding break the common commit merge. This might improve the developer experience for the first insight of pyspark. 

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Once the PR merged, users/developers can exec `dev/lint-python` for testing.

### CODE REVIEW

File: python/pyspark/ml/util.py [hunk 1]
```diff
@@ -536,10 +536,8 @@ def __get_class(clazz: str) -> Type[RL]:
         """
         parts = clazz.split(".")
         module = ".".join(parts[:-1])
-        m = __import__(module)
-        for comp in parts[1:]:
-            m = getattr(m, comp)
-        return m
+        m = __import__(module, fromlist=[parts[-1])
```
Comment (HyukjinKwon) [inline python/pyspark/ml/util.py:539]: I think it depends on which mypy version we use. cc @Yikun FYI.
If we can find a way to make all versions of mypy happy, that would be ideal.

**Comment by bzhaoopenstack** [comment]
`python/pyspark/streaming/context.py:372: error: Unused "type: ignore" comment
python/pyspark/streaming/context.py:378: error: Unused "type: ignore" comment
spark@DESKTOP-U0I7MO9:~/upstream/spark$ python --version
Python 3.8.13

CI:
Run PYTHON_EXECUTABLE=python3.9 ./dev/lint-python
  PYTHON_EXECUTABLE=python3.9 ./dev/lint-python
  shell: sh -e {0}
  env:
    LC_ALL: C.UTF-8
    LANG: C.UTF-8
    PYSPARK_DRIVER_PYTHON: python3.9
    PYSPARK_PYTHON: python3.9
    JAVA_HOME_8.0.33[2](https://github.com/bzhaoopenstack/spark/runs/7245194770?check_suite_focus=true#step:16:2)_x64: /__t/jdk/8.0.[3](https://github.com/bzhaoopenstack/spark/runs/7245194770?check_suite_focus=true#step:16:3)32/x6[4](https://github.com/bzhaoopenstack/spark/runs/7245194770?check_suite_focus=true#step:16:4)
    JAVA_HOME: /__t/jdk/8.0.332/x[6](https://github.com/bzhaoopenstack/spark/runs/7245194770?check_suite_focus=true#step:16:6)4
    JAVA_HOME_8_0_332_X64: /__t/jdk/8.0.332/x64
starting python compilation test...
python compilation succeeded.
starting black test...
black checks passed.
starting flake8 test...
flake8 checks passed.
starting mypy annotations test...
annotations failed mypy checks:
python/pyspark/streaming/context.py:3[7](https://github.com/bzhaoopenstack/spark/runs/7245194770?check_suite_focus=true#step:16:7)2: error: Argument 1 to "parallelize" of "SparkContext" has incompatible type "RDD[T]"; expected "Iterable[T]"  [arg-type]
python/pyspark/streaming/context.py:37[8](https://github.com/bzhaoopenstack/spark/runs/7245194770?check_suite_focus=true#step:16:8): error: Argument 1 to "parallelize" of "SparkContext" has incompatible type "RDD[T]"; expected "Iterable[T]"  [arg-type]
Found 2 errors in 1 file (checked 337 source files)
1
`

**Comment by Yikun** [comment]
@bzhaoopenstack Thanks for contributions!

1. Rename `[SPARK-39714][python]` to `[SPARK-39714][PYTHON]` : )
2. It would be good you can fullfill the PR description even this is a minor fix

```
./dev/lint-python
// ......
starting mypy annotations test...
annotations failed mypy checks:
python/pyspark/pandas/indexes/base.py:48: error: Module "pandas._libs" has no attribute "lib"  [attr-defined]
python/pyspark/streaming/context.py:372: error: Unused "type: ignore" comment
python/pyspark/streaming/context.py:378: error: Unused "type: ignore" comment
Found 3 errors in 2 files (checked 338 source files)
1
(venv) ➜  spark git:(pr/37117) mypy --version
mypy 0.941
```

In my env, maybe also fix these? Could you also take a look? lint CI are using [mypy==0.920](https://github.com/bzhaoopenstack/spark/runs/7246817397?check_suite_focus=true#step:9:22) I think

**Comment by bzhaoopenstack** [comment]
> @bzhaoopenstack Thanks for contributions!
> 
> 1. Rename `[SPARK-39714][python]` to `[SPARK-39714][PYTHON]` : )
> 2. It would be good you can fullfill the PR description even this is a minor fix
> 
> ```
> ./dev/lint-python
> // ......
> starting mypy annotations test...
> annotations failed mypy checks:
> python/pyspark/pandas/indexes/base.py:48: error: Module "pandas._libs" has no attribute "lib"  [attr-defined]
> python/pyspark/streaming/context.py:372: error: Unused "type: ignore" comment
> python/pyspark/streaming/context.py:378: error: Unused "type: ignore" comment
> Found 3 errors in 2 files (checked 338 source files)
> 1
> (venv) ➜  spark git:(pr/37117) mypy --version
> mypy 0.941
> ```
> 
> In my env, maybe also fix these? Could you also take a look? lint CI are using [mypy==0.920](https://github.com/bzhaoopenstack/spark/runs/7246817397?check_suite_focus=true#step:9:22) I think

Yes, my test result with mypy 0.920 with this PR is
```
spark@DESKTOP-U0I7MO9:~/upstream/spark$ dev/lint-python 
starting python compilation test...
python compilation succeeded.

The python3 -m black command was not found. Skipping black checks for now.

starting flake8 test...
flake8 checks passed.

starting mypy annotations test...
annotations passed mypy checks.

starting mypy examples test...
examples passed mypy checks.

starting mypy data test...
annotations passed data checks.


all lint-python tests passed!
spark@DESKTOP-U0I7MO9:~/upstream/spark$ mypy --version
mypy 0.920

```

File: python/pyspark/pandas/series.py [hunk 2]
```diff
@@ -4738,7 +4738,7 @@ def mode(self, dropna: bool = True) -> "Series":
         ser_count = self.value_counts(dropna=dropna, sort=False)
         sdf_count = ser_count._internal.spark_frame
         most_value = ser_count.max()
-        sdf_most_value = sdf_count.filter("count == {}".format(most_value))
+        sdf_most_value = sdf_count.filter("count == {}".format(most_value.__str__()))
```
Comment (HyukjinKwon) [inline python/pyspark/pandas/series.py:4741]:
```suggestion
        sdf_most_value = sdf_count.filter("count == {}".format(str(most_value)))
```

**Review by HyukjinKwon — APPROVED** [review]
Looks pretty good

**Comment by HyukjinKwon** [comment]
BTW, mind keeping the PR description template please? https://github.com/apache/spark/blob/master/.github/PULL_REQUEST_TEMPLATE

File: python/pyspark/ml/util.py (hunk 1 above)
Comment (Yikun) [reply python/pyspark/ml/util.py]: I also take a deep look on this:

> When the name variable is of the form package.module, normally, the top-level package (the name up till the first dot) is returned, not the module named by name. However, when a non-empty fromlist argument is given, the module named by name is returned.

such for `pyspark.ml.classification.LinearSVC`

```python
>>> clazz = "pyspark.ml.classification.LinearSVC";parts = clazz.split(".");module = ".".join(parts[:-1])
>>> __import__(module)
<module 'pyspark' from '/Users/yikun/spark/python/pyspark/__init__.py'>
>>> __import__(module, fromlist=[parts[-1]])
<module 'pyspark.ml.classification' from '/Users/yikun/spark/python/pyspark/ml/classification.py'>
```

and also return same results of wrong name like: `pyspark.ml.classification.unknow`, `pyspark.ml.unknow.LinearSVC`

After this, import with non-empty fromlist works well, can also simplified implementations and cleanup the mypy error:
```
python/pyspark/ml/util.py:542: error: Incompatible return value type (got Module, expected "Type[RL]")  [return-value]
```

So, I'm fine with this change.

File: python/pyspark/pandas/series.py (hunk 2 above)
Comment (bzhaoopenstack) [reply python/pyspark/pandas/series.py:4741]: Thanks, this is good. ;-)

**Comment by bzhaoopenstack** [comment]
> BTW, mind keeping the PR description template please? https://github.com/apache/spark/blob/master/.github/PULL_REQUEST_TEMPLATE

Thanks for correcting, sorry for that I'm not very familiar with Spark protocol. ;-) . Refreshed.

File: python/pyspark/ml/util.py [hunk 3]
```diff
@@ -536,10 +536,8 @@ def __get_class(clazz: str) -> Type[RL]:
         """
         parts = clazz.split(".")
         module = ".".join(parts[:-1])
-        m = __import__(module)
-        for comp in parts[1:]:
-            m = getattr(m, comp)
-        return m
+        m = __import__(module, fromlist=[parts[-1]])
+        return getattr(m, parts[-1])
```
Comment (HyukjinKwon) [inline python/pyspark/ml/util.py:540]: Hm, I think this changes the semantic. Previously, it returned the left-most import but not it returns right-most package. For example,

```python
>>> __import__("pyspark", fromlist=['sql'])
<module 'pyspark' from '/.../spark/python/pyspark/__init__.py'>
>>> m = __import__("pyspark")
>>> getattr(m, "sql")
<module 'pyspark.sql' from '/.../spark/python/pyspark/sql/__init__.py'>
```

Comment (Yikun) [reply python/pyspark/ml/util.py:540]:
```python
clazz = "pyspark"

parts = clazz.split(".")
module = ".".join(parts[:-1])
m = __import__(module)

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Empty module name

parts = clazz.split(".")
module = ".".join(parts[:-1])
m = __import__(module, fromlist=[parts[-1]])

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Empty module name
```

```python
clazz = "pyspark.sql"

# previous
parts = clazz.split(".")
module = ".".join(parts[:-1])
m = __import__(module)
for comp in parts[1:]:
    m = getattr(m, comp)

m
<module 'pyspark.sql' from '/Users/yikun/spark/python/pyspark/sql/__init__.py'>

# new
parts = clazz.split(".")
module = ".".join(parts[:-1])
m = __import__(module, fromlist=[parts[-1]])
getattr(m, parts[-1])
<module 'pyspark.sql' from '/Users/yikun/spark/python/pyspark/sql/__init__.py'>
```

BTW, the `from_list` value seems useless in our case. See also: https://stackoverflow.com/questions/2724260/why-does-pythons-import-require-fromlist

Comment (bzhaoopenstack) [reply python/pyspark/ml/util.py:540]: The original error from here is about type different error. As the previous function might return a python module type `m = __import__(module)`, then finally `return m` which can not be called, that's due to the [following line ](https://github.com/apache/spark/blob/master/python/pyspark/ml/util.py#L546). And the defined return-value is a Type[RL], they are different. So I think this func should return a python Class at least. That's why I change it.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**